src/System.Data.Common/src/System/Data/Common/DbDataReaderExtensions.cs

```diff
@@ -2,22 +2,92 @@
+using System.Collections.ObjectModel;
+
 namespace System.Data.Common
 {
     public static class DbDataReaderExtensions
     {
-        public static System.Collections.ObjectModel.ReadOnlyCollection<DbColumn> GetColumnSchema(this DbDataReader reader)
+        private sealed class DataRowDbColumn : DbColumn
+        {
+            private DataColumnCollection _schemaColumns;
+            private DataRow _schemaRow;
+
+            public DataRowDbColumn(DataRow readerSchemaRow, DataColumnCollection readerSchemaColumns)
+            {
+                _schemaRow = readerSchemaRow;
+                _schemaColumns = readerSchemaColumns;
+                PopulateFields();
+            }
+
+            private void PopulateFields()
+            {
+                AllowDBNull = GetDbColumnValue<bool?>(SchemaTableColumn.AllowDBNull);
+                BaseCatalogName = GetDbColumnValue<string>(SchemaTableOptionalColumn.BaseCatalogName);
+                BaseColumnName = GetDbColumnValue<string>(SchemaTableColumn.BaseColumnName);
+                BaseSchemaName = GetDbColumnValue<string>(SchemaTableColumn.BaseSchemaName);
+                BaseServerName = GetDbColumnValue<string>(SchemaTableOptionalColumn.BaseServerName);
+                BaseTableName = GetDbColumnValue<string>(SchemaTableColumn.BaseTableName);
+                ColumnName = GetDbColumnValue<string>(SchemaTableColumn.ColumnName);
+                ColumnOrdinal = GetDbColumnValue<int?>(SchemaTableColumn.ColumnOrdinal);
+                ColumnSize = GetDbColumnValue<int?>(SchemaTableColumn.ColumnSize);
+                IsAliased = GetDbColumnValue<bool?>(SchemaTableColumn.IsAliased);
+                IsAutoIncrement = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsAutoIncrement);
+                IsExpression = GetDbColumnValue<bool>(SchemaTableColumn.IsExpression);
+                IsHidden = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsHidden);
+                IsIdentity = GetDbColumnValue<bool?>("IsIdentity");
+                IsKey = GetDbColumnValue<bool?>(SchemaTableColumn.IsKey);
+                IsLong = GetDbColumnValue<bool?>(SchemaTableColumn.IsLong);
+                IsReadOnly = GetDbColumnValue<bool?>(SchemaTableOptionalColumn.IsReadOnly);
+                IsUnique = GetDbColumnValue<bool?>(SchemaTableColumn.IsUnique);
+                NumericPrecision = GetDbColumnValue<int?>(SchemaTableColumn.NumericPrecision);
+                NumericScale = GetDbColumnValue<int?>(SchemaTableColumn.NumericScale);
+                UdtAssemblyQualifiedName = GetDbColumnValue<string>("UdtAssemblyQualifiedName");
+                DataType = GetDbColumnValue<Type>(SchemaTableColumn.DataType);
+                DataTypeName = GetDbColumnValue<string>("DataTypeName");
+            }
+
+            private T GetDbColumnValue<T>(string columnName)
+            {
+                if (_schemaColumns.Contains(columnName) && _schemaRow[columnName] is T value)
```

Comment:
nit: use condition operator, convert method to expression-bodied.

Reply:
Similar to the above, it can be expression bodied but I think it's detrimental to readability if it is for this method. If don't feel as strongly as the switch expression though, again if you strongly prefer the expression version I can change it.

Reply:
In this case it does feel better, if you don't mind let's switch to the conditional.

Reply:
Ok, done.